rllib/core/learner/learner.py

```diff
@@ -1010,192 +1011,139 @@ def update_from_batch(
+                    del batch.policy_batches[module_id]
+            if not batch.policy_batches:
+                return {}
+
+            if minibatch_size:
+                batch_iter = MiniBatchCyclicIterator
+            elif num_epochs > 1:
+                # `minibatch_size` was not set but `num_epochs` > 1.
+                minibatch_size = batch.count
+                # Note that there is no need to shuffle here, b/c we don't have
+                # minibatches.
+                batch_iter = MiniBatchCyclicIterator
+            else:
+                # `minibatch_size` and `num_epochs` are not set by the user.
+                batch_iter = MiniBatchDummyIterator
 
-        if not self.iterator:
-            self.iterator = iterator
-
-        self._check_is_built()
+            batch = self._set_slicing_by_batch_id(batch, value=True)
 
-        # Call `before_gradient_based_update` to allow for non-gradient based
-        # preparations-, logging-, and update logic to happen.
-        self.before_gradient_based_update(timesteps=timesteps or {})
+            batch_iter = batch_iter(
+                batch,
+                num_epochs=num_epochs,
+                minibatch_size=minibatch_size,
+                shuffle_batch_per_epoch=shuffle_batch_per_epoch and (num_epochs > 1),
+                num_total_minibatches=num_total_minibatches,
+            )
 
-        def _finalize_fn(batch: Dict[str, numpy.ndarray]) -> Dict[str, Any]:
-            # Note, the incoming batch is a dictionary with a numpy array
-            # holding the `MultiAgentBatch`.
-            batch = self._convert_batch_type(batch["batch"][0])
-            return {"batch": self._set_slicing_by_batch_id(batch, value=True)}
-
-        i = 0
-        logger.debug(f"===> [Learner {id(self)}]: Looping through batches ... ")
-        while num_iters is None or i < num_iters:
-            for batch in self.iterator.iter_batches(
-                # Note, this needs to be one b/c data is already mapped to
-                # `MultiAgentBatch`es of `minibatch_size`.
-                batch_size=1,
-                _finalize_fn=_finalize_fn,
-                **kwargs,
-            ):
-                # TODO (simon): Add metrics for the `dataset_num_iter`.
-                # Update the iteration counter.
-                i += 1
-
-                # Note, `_finalize_fn`  must return a dictionary.
-                batch = batch["batch"]
-                logger.debug(
-                    f"===> [Learner {id(self)}]: batch {i} with {batch.env_steps()} rows."
-                )
-                # Check the MultiAgentBatch, whether our RLModule contains all ModuleIDs
-                # found in this batch. If not, throw an error.
-                unknown_module_ids = set(batch.policy_batches.keys()) - set(
-                    self.module.keys()
+        # Perform the actual looping through the minibatches or the given data iterator.
+        for tensor_minibatch in batch_iter:
```

Comment:
We need a counter here in case we iterate over an `DataIterator`. Maybe use `enumerate`.

Reply:
In case of a DataIterator, the new `MiniBatchRayDataIterator` will take care of the counting. It encapsulates all the logic of the old `update_from_iterator` method,. including the looping logic (and when to break).